### ISSUE TITLE
Reader View intermittently returns “Could not connect to server” error

### ISSUE THREAD

**Issue by cpeterso** [mozilla-mobile/firefox-ios#19855]
## Steps to reproduce

Starting a day or two ago in TestFlight Nightly, Reader View will be working and then sometimes stop working and return a “Could not connect to server” error. Then any attempt to use Reader View in that browsing session will also return that error. Reader View will sometimes start working again after killing and restarting the app, but not consistently.

Since this error is intermittent, it’s hard to bisect this regression. I installed Nightly builds from last week and I still experienced the error, but when I then updated back to the latest Nightly build, Reader View was working again. 

Could this error be related to crash https://github.com/mozilla-mobile/firefox-ios/issues/19783?

### Device & build information

* Device: iPhone 15 Plus running iOS 17.4.1
* Build version: 9000 (40735) but I see it in older builds too

### Notes

![image](https://github.com/mozilla-mobile/firefox-ios/assets/629193/bfb6646d-60cb-4a1f-bd98-c168635f4c2a)

**Comment by afurlan-firefox** [comment]
@cpeterso do you have different FX versions installed in the same device? It might collide in between the two.

**Comment by cpeterso** [comment]
> @cpeterso do you have different FX versions installed in the same device? It might collide in between the two. Can you restart the device, open just one instance and try again? Thanks!

Yes, I have installed Firefox from TestFlight and the App Store. Is Reader View implemented in a service outside of the Firefox app that both versions of Firefox would be using?? AFAIK, Firefox desktop and Android implement Reader View locally using the Readability.js library.

**Comment by clarmso** [comment]
I could reproduce what @cpeterso encountered using an iPhone 15 Plus simulator. I could not get out from the reader view by reloading the page or toggling the 📜 icon. A workaround is close the tab and open a new tab.
https://github.com/mozilla-mobile/firefox-ios/assets/1740517/b3c36c8e-9bd7-43b0-938a-23ef93ae5fc7

**Comment by cpeterso** [comment]
> Yes, I have two versions of Firefox installed: one from TestFlight and installed from the App Store. Is Reader View implemented in a service outside of the Firefox app that both versions of Firefox would be using?? 

Apparently this if the case. Steps to reproduce this bug:

1. Force quit both App Store Firefox and TestFlight Firefox, if they are running. I’m using App Store Firefox 125.1 and TestFlight Firefox 9000 (41001).
2. Launch TestFlight Firefox. 
3. Load a page and

**Comment by data-sync-user** [comment]
➤ Norberto Andres Furlan commented:

Nishant Bhasin is this the expected behavior?

**Comment by data-sync-user** [comment]
➤ Nishant Bhasin commented:

This is expected behaviour if you have two Firefox installed as at times it can happen where the port is left opened and the only way to close is it to either close both and if that also doesn’t fix then restart the device.

**Comment by cpeterso** [comment]
> This is expected behaviour if you have two Firefox installed as at times it can happen where the port is left opened

**Comment by cpeterso** [comment]
> This is expected behaviour if you have two Firefox installed as at times it can happen where the port is left opened

Why does Firefox Reader View need to communicate with a shared process instead of a library in process?

If communicating with a shared process is necessary and the port is left opened, that sounds like a Firefox resource leak. Does the Reader View port need to remain open after a page's Reader View layout has been rendered? Could there be a race condition between (for example) a user closing a Reader View tab before Firefox closes the Reader View port? Maybe we need a centralized port cleanup when the app is backgrounded?

**Comment by BaltimorePete** [comment]
I have this problem intermittently as well. I have Firefox Focus installed on my iPad, but it is never open and running when I have this issue.

**Comment by cpeterso** [comment]
@BaltimorePete What is your Firefox iOS version number? Firefox Focus doesn't have a reader view, so I don't think it would cause this issue. And as you said, you don't have it open when you have this issue.

**Comment by BaltimorePete** [comment]
Firefox 127.1 (42781)

**Comment by BaltimorePete** [comment]
Firefox 127.1 (42781)

**Comment by esthermations** [comment]
I just experienced this on Firefox 133.4 (49058), iOS 18.1.1. I only have one Firefox installed as far as I’m aware. I had Firefox Focus installed in the past (months ago) but not currently.

I was going to screen record the issue for this comment, so I closed my other tabs and swiped away all other apps in my app history just for privacy, but that actually seems to have fixed the issue. I didn’t even close Firefox. So maybe that’s helpful for anyone experiencing the issue.

**Comment by data-sync-user** [comment]
➤ Laurie Marceau commented:

I think this bug started happening following a bug fix for the reader mode. Either https://mozilla-hub.atlassian.net/browse/FXIOS-7681 ( https://mozilla-hub.atlassian.net/browse/FXIOS-7681|smart-link )  or https://mozilla-hub.atlassian.net/browse/FXIOS-7648 ( https://mozilla-hub.atlassian.net/browse/FXIOS-7648|smart-link )

**Comment by data-sync-user** [comment]
➤ Yoana Rios Diaz commented:

I tried to reproduce it but couldn’t. Andrei Bodea can you help find a way to reproduce this issue 🙏

**Comment by cpeterso** [comment]
I can still reproduce this error with App Store Firefox 135.0 (50174) and TestFlight Nightly 9000 (50908) using the STR in comment https://github.com/mozilla-mobile/firefox-ios/issues/19855#issuecomment-2068188273

**Comment by data-sync-user** [comment]
➤ Yoana Rios Diaz commented:

I can reproduce with 2 instances of Firefox but Nishant Bhasin mentioned here ( https://mozilla-hub.atlassian.net/browse/FXIOS-8998?focusedCommentId=881089 ) this is expected behaviour that’s why I was focusing on reproducing differently

**Comment by dimitarOnGithub** [comment]
Hi all,

I'm also running into this:

Phone: iPhone 13 Mini
OS: iOS 18.3.2
Firefox: App Store, version 136.3 (52294)

No secondary Firefox installation on my phone, only other browser I have installed is Safari.

Let me know if you'd like me to provide additional details.

**Comment by yoanarios** [comment]
Hi @dimitarOnGithub, Could you please provide the website and the steps to reproduce this issue? 🙏

**Comment by dimitarOnGithub** [comment]
Hi @yoanarios , of course, sorry about that!

I'm using reader mode to read through the book over here - https://www.cosmicpython.com/book/preface.html

Here are some pages where I got the same error as the one in the OP:
https://www.cosmicpython.com/book/part2.html
https://www.cosmicpython.com/book/chapter_08_events_and_message_bus.html
https://www.cosmicpython.com/book/chapter_11_external_events.html

I'd press the reader mode shortcut next to the URL and instead of loading the page in said mode, I get the `NSURLErrorDomain`; I tried on both wifi, as well as mobile data and a private tab,  but that had no effect whatsoever. Retrying in a few minutes usually works. Attaching two screenshots below.

![Image](https://github.com/user-attachments/assets/23031a27-599c-4e44-a4ee-21648ce5ad17)
![Image](https://github.com/user-attachments/assets/09aa92ba-41aa-4ba0-ae28-b2b5314352dd)

**Comment by data-sync-user** [comment]
➤ Matt P commented:

Andrei Bodea Can we check nightly to see if this issue is reproducible after recent reader mode changes?

**Comment by cpeterso** [comment]
I can still reproduce this error using Firefox 137.2 (53903) and TestFlight Nightly 9000 (54234).

**Comment by data-sync-user** [comment]
➤ Andrei Bodea commented:

Hello, I’m still able to reproduce this with Firefox v9000 (54305) with iPhone 15 Pro (18.5).

**Comment by data-sync-user** [comment]
➤ Andres Furlan commented:

Chris Peterson are you able to reproduce this in a production/release version? Same question for Andrei Bodea

**Comment by data-sync-user** [comment]
➤ Chris Peterson commented:

{quote}are you able to reproduce this in a production/release version?{quote}

Yes, but my STR require having two versions of Firefox installed. I have the production/release version from the App Store and the Nightly version from TestFlight. I don’t have STR to reproduce this error with only one version installed.

My STR:

  1. In the Nightly TestFlight version, open a page in reader mode.
  2. In the production/release version from the App Store, open a page in reader mode
  3. Then I get the “Could not connect to the server.” error.

**Comment by data-sync-user** [comment]
➤ Andres Furlan commented:

Chris Peterson perfect. Yes, we know that this happens when you have two versions. This is not a real issue in production. I am closing this ticket.
